theta_dup/include/theta_sketch_dup.h

```diff
@@ -0,0 +1,2209 @@
+   */
+  static bool hash_search_or_remove(uint64_t hash,
+                                    std::pair<uint64_t, uint64_t>* table,
+                                    uint8_t lg_size);
+  static bool hash_search(uint64_t hash,
+                          const std::pair<uint64_t, uint64_t>* table,
+                          uint8_t lg_size);
+
+  friend theta_sketch_dup_alloc<A>;
+  static update_theta_sketch_dup_alloc<A> internal_deserialize(
+      std::istream& is, resize_factor rf, uint8_t lg_cur_size,
+      uint8_t lg_nom_size, uint8_t flags_byte, uint64_t seed);
+  static update_theta_sketch_dup_alloc<A> internal_deserialize(
+      datasketches_pb::Update_theta_sketch_dup& pb, resize_factor rf,
+      uint8_t lg_cur_size, uint8_t lg_nom_size, uint8_t flags_byte,
+      uint64_t seed);
+  static update_theta_sketch_dup_alloc<A> internal_deserialize(
+      const void* bytes, size_t size, resize_factor rf, uint8_t lg_cur_size,
+      uint8_t lg_nom_size, uint8_t flags_byte, uint64_t seed);
+};
+
+// compact sketch
+
+template <typename A>
+class compact_theta_sketch_dup_alloc : public theta_sketch_dup_alloc<A> {
+ public:
+  static const uint8_t SKETCH_TYPE = 3;
+
+  // No constructor here.
+  // Instances of this type can be obtained:
+  // - by compacting an update_theta_sketch_dup
+  // - as a result of a set operation
+  // - by deserializing a previously serialized compact sketch
+
+  compact_theta_sketch_dup_alloc(const theta_sketch_dup_alloc<A>& other,
+                                 bool ordered);
+  virtual ~compact_theta_sketch_dup_alloc() = default;
+
+  virtual uint32_t get_num_retained() const;
+  virtual uint16_t get_seed_hash() const;
+  virtual bool is_ordered() const;
+  virtual string<A> to_string(bool print_items = false) const;
+  virtual void serialize(std::ostream& os) const;
+  virtual void serialize(datasketches_pb::Update_theta_sketch_dup* pb) const;
+  typedef vector_u8<A> vector_bytes;  // alias for users
+  // header space is reserved, but not initialized
+  virtual vector_bytes serialize(unsigned header_size_bytes = 0) const;
+
+  virtual typename theta_sketch_dup_alloc<A>::const_iterator begin() const;
+  virtual typename theta_sketch_dup_alloc<A>::const_iterator end() const;
+
+  /**
+   * This method deserializes a sketch from a given stream.
+   * @param is input stream
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of a sketch
+   */
+  static compact_theta_sketch_dup_alloc<A> deserialize(
+      std::istream& is, uint64_t seed = DEFAULT_SEED);
+
+  /**
+   * This method deserializes a sketch from protobuf.
+   * @param pb input stream
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of a sketch
+   */
+  static compact_theta_sketch_dup_alloc<A> deserialize(
+      datasketches_pb::Update_theta_sketch_dup& pb,
+      uint64_t seed = DEFAULT_SEED);
```

Comment:
Declaration: deserialize of another **derived class** compact_theta_sketch_dup_alloc using protobuf.